### ISSUE TITLE
DBus API: Issue configuring zone with policy

### ISSUE THREAD

**Issue by mheon** [firewalld/firewalld#874]
**What happened**:

I'm a Podman developer involved in implementing a new network stack for Podman. As part of this, we're implementing a direct firewalld interface to handle all firewall rules (unlike the existing solution which mixes direct iptables rules and firewalld) using policy options similar to [this blog](https://firewalld.org/2020/09/policy-objects-filtering-container-and-vm-traffic) but using the DBus interface directly instead of the `firewall-cmd` command line. However, firewalld is throwing a seemingly-unrelated error when I attempt to finalize configuration of the zone by adding the desired source subnet to it:

```
mcs@Bellerophon code/netavark (firewalld_impl *) $ sudo busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1 org.fedoraproject.FirewallD1.zone changeZoneOfSource ss netavark_zone 10.88.0.0/16                                                  
Call failed: COMMAND_FAILED: UNKNOWN_ERROR: nonexistent or underflow of priority count
```

I can't find this error in the docs and it doesn't seem to be associated at all with adding a source subnet.

I've included a reproducer using direct DBus calls via `busctl` to simplify the issue; the actual code Podman is executing is slightly more complicated to account for pre-existing zones and policies but follows the same general flow.

**What you expected to happen**:

Call succeeds, source subnet is successfully added to the zone.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a bridge interface on the host with name `podman0` and IP 10.88.0.1/16
2. Add a zone: `busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1/config org.fedoraproject.FirewallD1.config addZone2 sa{sv} testzone 0`
3. Add a policy associated with the zone (ingress zone testzone, egress ANY, masquerade enabled): `busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1/config org.fedoraproject.FirewallD1.config addPolicy sa{sv} testpolicy 3 egress_zones as 1 ANY ingress_zones as 1 testzone masquerade b true`
4. Reload the firewall: `busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1 org.fedoraproject.FirewallD1 reload`
5. Add the bridge interface to the zone: `busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1 org.fedoraproject.FirewallD1.zone changeZoneOfInterface ss netavark_zone podman0`
6. Add the source subnet to the zone: `busctl call org.fedoraproject.FirewallD1 /org/fedoraproject/FirewallD1 org.fedoraproject.FirewallD1.zone changeZoneOfSource ss netavark_zone 10.88.0.0/16`

**Anything else we need to know?**:

After this, subsequent `firewall-cmd --reload` invocations fail with inconsistent `python-nftables` errors; a `systemctl restart firewalld` is necessary to fix this.

**Environment**:

Stock Fedora 34 installation, all commands run as root.

- Firewalld Version (if Fedora based `dnf info firewalld` or commit hash if developing from git `git log -n1 --format=format:"%H"`):

firewalld-0.9.4-1.fc34.noarch

- Firewalld Backend (`cat /etc/firewalld/firewalld.conf | grep FirewallBackend`): 

nftables

- OS (e.g: `cat /etc/os-release`):

Fedora 34

- Others:

**Comment by erig0** [comment]
Having an interface in the ingress zone while using masquerade is invalid. This is because nftables/iptables can't match the ingress interface (iifname) in the postrouting hooks. Can you try without the interface?

Unfortunately firewalld is not detecting this in your config due to a different bug. The config checks really need some love.

**Comment by mheon** [comment]
Roger, thank you - will check this morning and report back.

**Comment by mheon** [comment]
Can confirm that resolved the configuration issue, though it looks like traffic either is not being masqeuraded or is otherwise being dropped before egress, attempting to debug now.

**Comment by mheon** [comment]
Appears that traffic is being dropped before it gets to the public interface - I don't see any traffic emerged, masqueraded or no, on the public interface. @erig0 Is there any easy way to debug this further? I could look into the nftables rules firewalld is producing but they're a bit labyrinthine.

**Comment by rhatdan** [comment]
This is holding us up. Could we get some help on this.

**Comment by erig0** [comment]
Matt and I talked offline. Copy/pasting it here.

tl;dr masquerade was enabled, but you still need to allow/accept the traffic.

---

> Got past the initial hurdle - firewalld now accepts the config - but it   
> looks like the traffic is being dropped prior to making it to the outbound
> interface (per Wireshark). What's the best way to debug this - directly   
> looking at the nftables rules?                                            
                                                                            
Start with firewalld, enable --set-log-denied (persistent config            
change). That should tell you where it's dropping.                          
```                                                                              
  # firewall-cmd --set-log-denied=all                                       
```                                                                              
By default policies use --set-target=CONTINUE. If all you did was enable    
masquerade in the policy then you likely also need to allow/accept the      
traffic.                                                                    
                                                                            
To allow _all_ traffic, change --set-target:                                
```                                                                            
  # firewall-cmd --permanent --policy <policy> --set-target=ACCEPT          
```                                                                              
Alternatively you can accept a subset of traffic. For example to only       
masquerade https traffic accept only https:                                 
```                                                                              
  # firewall-cmd --permanent --policy <policy> --service https              
```

**Comment by erig0** [comment]
@mheon, can this be closed? The config checks are being worked on outside of this bug.

**Comment by mheon** [comment]
Sure

**Comment by erig0** [comment]
PSA: I think #910 catches this configuration issue.

https://github.com/firewalld/firewalld/pull/910/commits/44e61b5a1095d8370e3dffa696c071c62c62e1d7#diff-81ef949dd8a17eabe7aa55951a339b920bc6560b11b5d219fc36ff104dde5bf5
